dali/imgcodec/image_decoder.cc

```diff
@@ -0,0 +1,607 @@
+  return ScheduleDecode(ctx, make_span(&out, 1), make_span(&in, 1), opts, make_span(&roi, 1));
+}
+
+FutureDecodeResults ImageDecoder::ScheduleDecode(DecodeContext ctx,
+                                                span<SampleView<CPUBackend>> out,
+                                                cspan<ImageSource *> in,
+                                                DecodeParams opts,
+                                                cspan<ROI> rois) {
+  temp_buffers_.clear();
+  int N = out.size();
+  assert(in.size() == N);
+  assert(rois.size() == N || rois.empty());
+  DecodeResultsPromise results(N);
+
+  auto work = new_work(ctx, results);
+  work->cpu_outputs.reserve(N);
+  work->indices.reserve(N);
+  work->sources.reserve(N);
+
+  for (auto &o : out)
+    work->cpu_outputs.push_back(o);
+  for (auto *src : in)
+    work->sources.push_back(src);
+  for (int i = 0; i < N; i++)
+    work->indices.push_back(i);
+
+  DistributeWork(std::move(work));
+
+  temp_buffers_.clear();
+
+  return results.get_future();
+}
+
+
+/**
+ * @brief Decodes a single image to device buffers
+ */
+FutureDecodeResults ImageDecoder::ScheduleDecode(DecodeContext ctx,
+                                                 span<SampleView<GPUBackend>> out,
+                                                 cspan<ImageSource *> in,
+                                                 DecodeParams opts,
+                                                 cspan<ROI> rois) {
+  int N = out.size();
+  assert(in.size() == N);
+  assert(rois.size() == N || rois.empty());
+
+  temp_buffers_.resize(N);
+
+  DecodeResultsPromise results(N);
+
+  auto work = new_work(ctx, results);
+  work->gpu_outputs.reserve(N);
+  work->indices.reserve(N);
+  work->sources.reserve(N);
+
+  for (auto &o : out)
+    work->gpu_outputs.push_back(o);
+  for (auto *src : in)
+    work->sources.push_back(src);
+  for (int i = 0; i < N; i++)
+    work->indices.push_back(i);
+
+  DistributeWork(std::move(work));
+
+  return results.get_future();
+}
+
+void ImageDecoder::DistributeWork(std::unique_ptr<ScheduledWork> work) {
+  std::map<const ImageFormat*, std::unique_ptr<ScheduledWork>> dist;
+  for (int i = 0; i < work->num_samples(); i++) {
+    auto *f = FormatRegistry().GetImageFormat(work->sources[i]);
+    if (!f || filtered_.find(f) == filtered_.end()) {
+      work->results.set(i, DecodeResult::Failure(
+        std::make_exception_ptr(std::runtime_error("Image not supported."))));
+      continue;
+    }
+    auto &w = dist[f];
+    if (!w)
+      w = new_work(work->ctx, work->results);
+    w->move_entry(*work, i);
+  }
+  for (auto &[f, w] : dist) {
+    auto &worker = workers_[filtered_.find(f)->second];
+    worker->add_work(std::move(w));
+  }
+}
+
+void ImageDecoder::InitWorkers(bool lazy_init) {
+  const ImageFormat *prev_format = nullptr;
+  DecoderWorker *prev_worker = nullptr;
+  for (auto [format, factory] : filtered_) {
```

Comment:
Why do we have DecoderWorker active objects per format; rather than simply use thread_pool of configurable size and schedule work to be done there?

Reply:
There are multiple reasons:
1. The workers schedule work for one another. The decoder fallback system is a DAG, so there's no risk of deadlock. While the number of decoders is fixed, the number of work items in the queues is not guaranteed not to exceed the number of decoders (pessimistically it can reach the batch size). In that case, it's possible that a cycle would form.
2. The work is always the same - only the data changes, but it's always the same function. only the data change. The data objects are kept in a pool to avoid extra allocations.
3. ScheduleDecode in `decoder_` is not thread-safe. Here, there's one dedicated thread that calls it.

I'm not saying that this work scheme could not be abstracted - it could, but that'd come with additional development time cost which we cannot afford at this point.

Reply:
What's the point of keeping temporary buffers alive here instead of relying on the memory resource?

Reply:
@jantonguirao I think this comment is misplaced.

Reply:
It was a general question, following your explanation above